File: ros2_ws/install/robovision_services/lib/robovision_services/robovision_service.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from geometry_msgs.msg import Pose
from sensor_msgs.msg import Image, PointCloud2
import sensor_msgs_py.point_cloud2 as pc2

import cv2
from cv_bridge import CvBridge

import numpy as np
import traceback

from robovision_interfaces.msg import ObjectCentroid
from robovision_interfaces.srv import GetPointCenter

class PointCloudCentroidNode(Node):
    def __init__(self):
        super().__init__("point_cloud_service")

        self.rgb_ = []
        self.depth_ = []
        self.depth_mat_ = []
        self.point_cloud_ = []

        self.centroid_=Pose()

        self.is_ptcld_ = False
        self.display_ = True

        #Publishers
        #self.centroid_publisher_ = self.create_publisher(
        #    Pose, "/object_centroid", 10)

        #Subscribers
        self.subscriber_ = self.create_subscription(
            Image, "/camera/rgb/image_rect_color", self.callback_rgb_rect, 10)
        self.subscriber_ = self.create_subscription(
            Image, "/camera/depth_registered/image", self.callback_depth_rect, 10)
        self.subscriber_ = self.create_subscription(
            PointCloud2, "/camera/depth_registered/points", self.callback_point_cloud, 10)
        
        #Processing
        #self.processing_timer_ = self.create_timer(0.030, self.point_cloud_processing) #update image each 30 miliseconds

        #Services
        self.get_point_center_service_ = self.create_service(
            GetPointCenter, "get_point_center", self.point_cloud_processing)
        
        self.get_logger().info("Starting point_cloud_service application in python...")

    def callback_rgb_rect(self, msg):
        try:
            #Transform the new message to an OpenCV matrix
            bridge_rgb=CvBridge()
            self.rgb_ = bridge_rgb.imgmsg_to_cv2(msg,msg.encoding).copy()
            self.rgb_ = cv2.cvtColor(self.rgb_, cv2.COLOR_RGB2BGR)

        except:
            self.get_logger().error(traceback.format_exc())

    def callback_depth_rect(self, msg):
        try:
            #Transform the new message to an OpenCV matrix
            bridge_depth=CvBridge()
            self.depth_=bridge_depth.imgmsg_to_cv2(msg,"32FC1").copy()

            self.depth_mat_ = np.array(self.depth_, dtype=np.float32)
            cv2.normalize(self.depth_mat_, self.depth_, 0, 1, cv2.NORM_MINMAX)

        except:
            self.get_logger().error(traceback.format_exc())

    def callback_point_cloud(self, msg):
        try:
            # Read the point cloud as a list of tuples (x, y, z, ...)
            self.point_cloud_ = np.array(list(pc2.read_points(msg, field_names=None, skip_nans=False)))

            # Reshape if the point cloud is organized (msg.height > 1)
            if msg.height > 1:
                self.point_cloud_ = self.point_cloud_.reshape((msg.height, msg.width, -1))

            rows, cols, _ = self.point_cloud_.shape
            self.get_logger().info('Point cloud received with size: rows: {}, cols: {}'.format(rows, cols))

            self.is_ptcld_ = True
        except:
            self.get_logger().error(traceback.format_exc())

    def point_cloud_processing(self, request, response):
        
        response.point = ObjectCentroid()

        response.point.x = 0.0
        response.point.y = 0.0
        response.point.z = 0.0
        response.point.centroid = [0.0, 0.0, 0.0]

        if self.is_ptcld_:
            if request.x and request.y:
                #Access a point in the point_cloud
                col_id = request.x
                row_id = request.y

                p = [float(self.point_cloud_[row_id, col_id, 0][0]), 
                    float(self.point_cloud_[row_id, col_id, 0][1]), 
                    float(self.point_cloud_[row_id, col_id, 0][2])]

                response.point.x = p[0]
                response.point.y = p[1]
                response.point.z = p[2]
                response.point.centroid = p


                self.get_logger().info("Centroid at ({}, {}): [{}, {}, {}]".format(row_id, col_id, p[0], p[1], p[2]))

        else:
            self.get_logger().error("No PointCloud data available for processing.")

        #Publish centroid pose
        #self.centroid_publisher_.publish(self.centroid_)

        return response


def main(args=None):
    rclpy.init(args=args)
    node = PointCloudCentroidNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        cv2.destroyAllWindows()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/robovision_ros2/robovision_4/src/robovision_service.cpp
#include "rclcpp/rclcpp.hpp"

#include "geometry_msgs/msg/pose.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "sensor_msgs/msg/point_field.hpp"

#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

#include "robovision_interfaces/srv/get_point_center.hpp"
#include "robovision_interfaces/msg/object_centroid.hpp"

#include <string>
#include <vector>


class PointCloudCentroidNode : public rclcpp::Node
{
public:
    PointCloudCentroidNode() : Node("point_cloud_service"), is_ptcld_(false), display_(true)
    {
        // Subscribers
        rgb_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera/rgb/image_rect_color", 10,
            std::bind(&PointCloudCentroidNode::callback_rgb_rect, this, std::placeholders::_1));

        depth_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera/depth_registered/image", 10,
            std::bind(&PointCloudCentroidNode::callback_depth_rect, this, std::placeholders::_1));

        point_cloud_sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
            "/camera/depth_registered/points", 10,
            std::bind(&PointCloudCentroidNode::callback_point_cloud, this, std::placeholders::_1));

        // Services
        get_point_center_service_ = this->create_service<robovision_interfaces::srv::GetPointCenter>(
            "get_point_center", 
            std::bind(&PointCloudCentroidNode::point_cloud_processing, this, 
                      std::placeholders::_1, std::placeholders::_2));

        RCLCPP_INFO(this->get_logger(), "Starting point_cloud_service application in cpp...");
    }

private:

    bool is_ptcld_;
    bool display_;

    cv::Mat rgb_, depth_, depth_mat_, point_cloud_;
    
    geometry_msgs::msg::Pose centroid_;

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr rgb_sub_;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr depth_sub_;
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr point_cloud_sub_;
    
    rclcpp::Service<robovision_interfaces::srv::GetPointCenter>::SharedPtr get_point_center_service_;

    void callback_rgb_rect(const sensor_msgs::msg::Image::SharedPtr msg) 
    {
        try 
        {
            cv_bridge::CvImagePtr bridge_rgb = cv_bridge::toCvCopy(msg, msg->encoding);
            rgb_ = bridge_rgb->image.clone();
            cv::cvtColor(rgb_, rgb_, cv::COLOR_RGB2BGR);
        }
        catch (const std::exception &e) 
        {
            RCLCPP_ERROR(this->get_logger(), "Error processing RGB image: %s", e.what());
        }
    }

    void callback_depth_rect(const sensor_msgs::msg::Image::SharedPtr msg) 
    {
        try 
        {
            cv_bridge::CvImagePtr bridge_depth = cv_bridge::toCvCopy(msg, "32FC1");
            depth_ = bridge_depth->image.clone();

            depth_mat_ = depth_.clone();
            depth_.convertTo(depth_mat_, CV_32F);
            cv::normalize(depth_mat_, depth_, 0, 1, cv::NORM_MINMAX);
        } 
        catch (const std::exception &e) 
        {
            RCLCPP_ERROR(this->get_logger(), "Error processing Depth image: %s", e.what());
        }
    }

    void callback_point_cloud(const sensor_msgs::msg::PointCloud2::SharedPtr msg) 
    {
        try 
        {
            // Ensure the point cloud is organized
            if (msg->height > 1)
            {
                // Convert PointCloud2 to cv::Mat
                point_cloud_ = cv::Mat(msg->height, msg->width, CV_32FC4);
                const uint8_t *data_ptr = msg->data.data();

                for (size_t i = 0; i < msg->height; ++i) 
                {
                    for (size_t j = 0; j < msg->width; ++j) 
                    {
                        const float *point = reinterpret_cast<const float*>(data_ptr + (i * msg->row_step + j * msg->point_step));
                        point_cloud_.at<cv::Vec4f>(i, j) = cv::Vec4f(point[0], point[1], point[2], point[3]);
                    }
                }
                
                is_ptcld_ = true;
                RCLCPP_INFO(this->get_logger(), "Point cloud received with size: %d x %d", msg->height, msg->width);
            }
            else
            {
                RCLCPP_WARN(this->get_logger(), "Point cloud is not organized. Cannot convert to cv::Mat.");
            }
        } 
        catch (const std::exception &e) 
        {
            RCLCPP_ERROR(this->get_logger(), "Error processing PointCloud: %s", e.what());
        }
    }

    void point_cloud_processing(
        const std::shared_ptr<robovision_interfaces::srv::GetPointCenter::Request> request,
        std::shared_ptr<robovision_interfaces::srv::GetPointCenter::Response> response) 
    {
        response->point.x = 0.0;
        response->point.y = 0.0;
        response->point.z = 0.0;
        response->point.centroid = {0.0, 0.0, 0.0};

        if (!is_ptcld_) 
        {
            RCLCPP_ERROR(this->get_logger(), "No PointCloud data available for processing.");
            return;
        }

        try 
        {
            int row_id = request->y;
            int col_id = request->x;

            if (row_id >= 0 && row_id < point_cloud_.rows && col_id >= 0 && col_id < point_cloud_.cols) 
            {
                cv::Vec4f point = point_cloud_.at<cv::Vec4f>(row_id, col_id);

                response->point.x = point[0];
                response->point.y = point[1];
                response->point.z = point[2];
                response->point.centroid = {point[0], point[1], point[2]};

                RCLCPP_INFO(this->get_logger(), "Centroid at (%d, %d): [%f, %f, %f]",
                            row_id, col_id, point[0], point[1], point[2]);
            } 
            else 
            {
                RCLCPP_ERROR(this->get_logger(), "Requested coordinates are out of bounds.");
            }
        } 
        catch (const std::exception &e) 
        {
            RCLCPP_ERROR(this->get_logger(), "Error in processing PointCloud: %s", e.what());
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PointCloudCentroidNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
